Rainjacked/rainscript
# Tokenizer

# Summary

This is a proposal to use a regex-based lexer for tokenizing in Rainscript.

The ideal usage of the tokenizer is a is to generate a minimal set of tokens (axioms) for unit patterns, such as comments, string literals, integer literals, keyword patterns, and multi-character operators. The output of the tokenizer will become the input for the parser.

## Tokenizer Rules

**Event**
- [X] EVENT - the event name 

**Variables**
- [X] VAR_SUFFIX - possible suffix names variables

**Literals**
- [x] STR_2 - double-quoted string literal
- [x] STR_1 - single-quoted string literal
- [x] FLOAT - floating-point number literal
- [X] INT_OCT - octal integer literal
- [X] INT_DEC - decimal integer literal
- [X] INT_BIN - binary integer literal
- [X] INT_HEX - hexadecimal integer literal

**Miscellaneous**
- [x] COMMENT - inline- or block-comment

# How to Test

1. Have a fresh clone of this PR by running:
```
git clone https://github.com/Hikari9/rainscript -b tokenizer rainscript-tokenizer
cd rainscript-tokenizer
```
  Alternatively, you may also fetch using a remote to `Hikari9/parser-00-util`.

2. For first use, make sure npm is installed then run:
```
npm install
```

3. To perform tests on your machine, run:
```
npm test
```

4. (Optional) Add your own tests to `/tests/tokenizer/`. Push the new tests into this PR so that it can be reviewed (if you are a maintainer). Otherwise simply leave a comment.
